File: web-pixi/src/clockSync.ts
// Window of recent `instant = serverTime - clientNow` observations used
// to derive offsetMs. 40 frames @ 20Hz server tick ≈ 2s of history —
// enough to absorb the bursty-delivery pattern seen on the WSL2 path
// (multi-frame bursts every ~300ms) without going so long that a real
// shift in base latency takes forever to track.
const INSTANT_WINDOW = 40;

/**
 * ClockSync derives the server-to-client wall-clock offset from per-frame
 * `producedAtMs` observations.
 *
 * Algorithm: **sliding-window max of `instant = serverTime - clientNow`**.
 *
 * Why max instead of EWMA: bursty WebSocket delivery (OS-level coalescing,
 * main-thread stalls queueing onmessage events, etc.) produces clusters of
 * frames where clientNow stays fixed while serverStamp ticks forward —
 * each frame in the cluster has a progressively LESS NEGATIVE instant.
 * Once normal cadence resumes, instant stabilizes near the value of the
 * LAST burst frame. The earlier burst frames carry pure delay noise and
 * are pathological for an EWMA — they pull offsetMs hundreds of ms away
 * from the truth.
 *
 * The maximum (least negative) instant in the recent window corresponds
 * to the frame that arrived with the LEAST delay — the closest sample to
 * a true zero-buffering reading. Sliding-window scope (~2s) lets the
 * offset drift if base latency genuinely shifts, by aging out the old
 * max sample.
 */
export interface ClockSync {
  /** Best estimate of server_ms − client_ms offset. */
  offsetMs: number;
  /** True once at least one frame has been observed. */
  initialized: boolean;
  /** Ring of recent instant observations. */
  instants: Float64Array;
  instantsIdx: number;
  instantsCount: number;
}

export function newClockSync(): ClockSync {
  return {
    offsetMs: 0,
    initialized: false,
    instants: new Float64Array(INSTANT_WINDOW),
    instantsIdx: 0,
    instantsCount: 0,
  };
}

/** Minimal shape required to anchor a frame on the clock. */
interface HasProducedAt { producedAtMs: number; }

/**
 * Feed one server timestamp observation with the client's
 * performance.now() at the moment of observation. Pushes into the
 * sliding window and recomputes offsetMs as the window's max.
 */
export function observeServerTime(
  c: ClockSync,
  serverTimeMs: number,
  clientNowMs: number,
): void {
  const instant = serverTimeMs - clientNowMs;

  c.instants[c.instantsIdx] = instant;
  c.instantsIdx = (c.instantsIdx + 1) % c.instants.length;
  if (c.instantsCount < c.instants.length) c.instantsCount++;

  if (!c.initialized) {
    c.offsetMs = instant;
    c.initialized = true;
    return;
  }

  // Max over the populated portion of the ring. O(40) per frame is
  // negligible (called per server frame ≈ 20Hz).
  let maxInstant = -Infinity;
  for (let i = 0; i < c.instantsCount; i++) {
    if (c.instants[i] > maxInstant) maxInstant = c.instants[i];
  }
  c.offsetMs = maxInstant;
}

/**
 * Feed the clock the newest `producedAtMs` across a frame's decoded
 * entities. Every entity in a frame carries its own ClusterClock-aligned
 * stamp; we use the freshest one as this frame's observation. No-ops if
 * the frame contains no entities with stamps.
 */
export function observeFrameStamps(
  c: ClockSync,
  entities: readonly HasProducedAt[],
  clientNowMs: number,
): void {
  let maxStamp = 0;
  for (const e of entities) {
    if (e.producedAtMs > maxStamp) maxStamp = e.producedAtMs;
  }
  if (maxStamp > 0) observeServerTime(c, maxStamp, clientNowMs);
}

/** Estimated current server wall-clock time in ms, given a client performance.now() reading. */
export function estimatedServerNow(c: ClockSync, clientNowMs: number): number {
  return clientNowMs + c.offsetMs;
}

File: csharp/Mmokit.Sdk.Core/ClockSync.cs
using System.Collections.Generic;

namespace Mmokit.Sdk.Core
{
    /// Server-to-client wall-clock offset estimator. Faithful port of
    /// pkg/quantize/ts/clock-sync.ts: offset = sliding-window MAX of
    /// instant = serverMs - clientNowMs over the last InstantWindow samples.
    /// Max (not EWMA) because bursty delivery clusters frames with a fixed
    /// clientNow and advancing serverStamp; the max is the least-delayed
    /// (truest) reading, and window scope lets it drift if base latency shifts.
    ///
    /// Pure + engine-agnostic (clock injected as a scalar). The CONSUMER drives
    /// it from decoded producedAtMs stamps — consistent with the stateless SDK.
    public sealed class ClockSync
    {
        public const int InstantWindow = 40;

        readonly object _sync = new();
        readonly double[] _instants = new double[InstantWindow];
        int _idx;
        int _count;
        double _offsetMs;
        bool _initialized;

        public double OffsetMs
        {
            get
            {
                lock (_sync)
                    return _offsetMs;
            }
        }

        public bool Initialized
        {
            get
            {
                lock (_sync)
                    return _initialized;
            }
        }

        /// Feed one (serverTimeMs, clientNowMs) observation; recompute offset.
        public void ObserveServerTime(double serverTimeMs, double clientNowMs)
            => ObserveServerTimeAndGetOffset(serverTimeMs, clientNowMs);

        /// Observe and return the resulting offset under one clock lock. This
        /// lets AdaptivePlaybackController use the exact observation it just
        /// submitted even when the clock is shared by multiple controllers.
        internal double ObserveServerTimeAndGetOffset(double serverTimeMs, double clientNowMs)
        {
            lock (_sync)
            {
                return ObserveServerTimeUnsafe(serverTimeMs, clientNowMs);
            }
        }

        /// Atomically replace a superseded producer's samples and observe the
        /// replacement stream's first timestamp.
        internal double ResetAndObserveServerTimeAndGetOffset(double serverTimeMs, double clientNowMs)
        {
            lock (_sync)
            {
                ResetUnsafe();
                return ObserveServerTimeUnsafe(serverTimeMs, clientNowMs);
            }
        }

        /// Feed the freshest producedAtMs across a frame's decoded entities.
        public void ObserveFrameStamps(IEnumerable<ulong> producedAtMs, double clientNowMs)
        {
            ulong maxStamp = 0;
            foreach (var p in producedAtMs)
                if (p > maxStamp) maxStamp = p;
            if (maxStamp > 0) ObserveServerTime(maxStamp, clientNowMs);
        }

        /// Drop samples from a superseded producer stream. A replacement
        /// authority may legitimately publish an earlier cluster-clock stamp.
        public void Reset()
        {
            lock (_sync)
            {
                ResetUnsafe();
            }
        }

        double ObserveServerTimeUnsafe(double serverTimeMs, double clientNowMs)
        {
            double instant = serverTimeMs - clientNowMs;
            _instants[_idx] = instant;
            _idx = (_idx + 1) % _instants.Length;
            if (_count < _instants.Length) _count++;

            if (!_initialized)
            {
                _offsetMs = instant;
                _initialized = true;
                return _offsetMs;
            }
            double max = double.NegativeInfinity;
            for (int i = 0; i < _count; i++)
                if (_instants[i] > max) max = _instants[i];
            _offsetMs = max;
            return _offsetMs;
        }

        void ResetUnsafe()
        {
            System.Array.Clear(_instants, 0, _instants.Length);
            _idx = 0;
            _count = 0;
            _offsetMs = 0;
            _initialized = false;
        }

        /// Estimated current server wall-clock ms, given a client clock reading.
        public double EstimatedServerNow(double clientNowMs)
        {
            lock (_sync)
                return clientNowMs + _offsetMs;
        }

        /// Atomically test initialization and estimate server time.
        internal bool TryEstimatedServerNow(double clientNowMs, out double serverNowMs)
        {
            lock (_sync)
            {
                serverNowMs = clientNowMs + _offsetMs;
                return _initialized;
            }
        }
    }
}
